### PR TITLE
Manage PluginFactory plugins with unique_ptr in RecoLocalMuon

### PR DESCRIPTION
**Pull request by makortel**
Changed also some other raw pointers to `unique_ptr`, and used `unique_ptr` constructor in `RPCRecHitProducer`. This PR is preparatory work to change the PluginFactory to return a `std::unique_ptr`.

Tested in CMSSW_10_5_X_2019-02-05-1100 , no changes expected.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

File: RecoLocalMuon/DTRecHit/plugins/DTRecHitProducer.cc [hunk 1]
```diff
@@ -32,25 +32,22 @@ using namespace std;
...
-
-  // Get the concrete reconstruction algo from the factory
-  string theAlgoName = config.getParameter<string>("recAlgo");
-  theAlgo = DTRecHitAlgoFactory::get()->create(theAlgoName,
-					       config.getParameter<ParameterSet>("recAlgoConfig"));
 }
 
 DTRecHitProducer::~DTRecHitProducer(){
   if(debug)
     cout << "[DTRecHitProducer] Destructor called" << endl;
```
[the remark this answers is not in the document]
Comment (perrotta) [reply RecoLocalMuon/DTRecHit/plugins/DTRecHitProducer.cc]: I'd say that none of these debug printouts are relevant, at least not the ones in the c'tor and d'tor, but maybe also the one in the main method which just counts the number of hits in an unspecified DT layer: could it be simplified further by just removing the "debug" parameter and all its calls in this producer?

@namapane @fcavallo please let us know (answering if you still need the debug printout in the destructor at first, and all the other ones as second)

File: RecoLocalMuon/DTSegment/src/DTRecSegment2DProducer.cc [hunk 2]
```diff
@@ -46,17 +49,13 @@ DTRecSegment2DProducer::DTRecSegment2DProducer(const edm::ParameterSet& pset) {
...
-  if(debug) cout << "the Reco2D AlgoName is " << theAlgoName << endl;
-  theAlgo = DTRecSegment2DAlgoFactory::get()->create(theAlgoName,
-                                                     pset.getParameter<ParameterSet>("Reco2DAlgoConfig"));
+  if(debug) cout << "the Reco2D AlgoName is " << pset.getParameter<string>("Reco2DAlgoName") << endl;
 }
 
 /// Destructor
 DTRecSegment2DProducer::~DTRecSegment2DProducer() {
   if(debug)
     cout << "[DTRecSegment2DProducer] Destructor called" << endl;
```
[the remark this answers is not in the document]
Comment (perrotta) [reply RecoLocalMuon/DTSegment/src/DTRecSegment2DProducer.cc]: @namapane @fcavallo : please answer

File: RecoLocalMuon/DTRecHit/plugins/DTRecHitProducer.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (namapane) [reply RecoLocalMuon/DTRecHit/plugins/DTRecHitProducer.cc]: I agree the printouts can be removed; I don't think the debug option has ever been activated in the last 10 years at least. 
Whether you want to remove all printouts  or only those in the dtor is your choice, I generally prefer not to touch working code unless a fix is needed, but if you prefer to clean up I don't have any objection.

[the remark this answers is not in the document]
Comment (makortel) [reply RecoLocalMuon/DTRecHit/plugins/DTRecHitProducer.cc]: @perrotta Which way do you prefer?

[the remark this answers is not in the document]
Comment (slava77) [reply RecoLocalMuon/DTRecHit/plugins/DTRecHitProducer.cc]: if this does not resolve any code check/quality analysis issue, perhaps we just leave it as is.

[the remark this answers is not in the document]
Comment (makortel) [reply RecoLocalMuon/DTRecHit/plugins/DTRecHitProducer.cc]: > if this does not resolve any code check/quality analysis issue, perhaps we just leave it as is.

I don't think we flag `cout`s, so it indeed would not. I'm fine with leaving this as they are as well.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/25899	2a951b3
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no (relevant) differences

**Comment by kpedro88** [comment]
+upgrade

**Comment by kpedro88** [comment]
+1

**Pull request merged**